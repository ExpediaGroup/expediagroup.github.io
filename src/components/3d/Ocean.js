/*
Copyright 2022 Expedia, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as THREE from 'three'
import React, { useRef, useMemo } from 'react'
import { extend, useThree, useLoader, useFrame } from '@react-three/fiber'
import { Water } from 'three/examples/jsm/objects/Water'

extend({ Water })

/**
 * Plane with water effect.
 */
function Ocean({sunDirection = new THREE.Vector3()}) {
  const thisMesh = useRef()
  const gl = useThree((state) => state.gl)
  const waterNormals = useLoader(THREE.TextureLoader, '/img/waternormals.jpg')
  waterNormals.wrapS = waterNormals.wrapT = THREE.RepeatWrapping
  const geom = useMemo(() => new THREE.PlaneGeometry(10000, 10000), [])
  const config = useMemo(
    () => ({
      textureWidth: 512,
      textureHeight: 512,
      waterNormals,
      sunDirection: sunDirection,
      sunColor: 0xffffff,
      waterColor: 0x18b3a3,
      distortionScale: 1,
      fog: false,
      format: gl.encoding
    }),
    [waterNormals]
  )
  useFrame((state, delta) => (thisMesh.current.material.uniforms.time.value += delta / 3))
  return <water ref={thisMesh} args={[geom, config]} rotation-x={- Math.PI / 2} />
}

export default Ocean;
