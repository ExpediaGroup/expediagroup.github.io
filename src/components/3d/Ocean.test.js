/*
Copyright 2022 Expedia, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

jest.mock('three/examples/jsm/objects/Water', () => {
    return { Water: 'Water' }
})
jest.mock('@react-three/fiber', () => {
    return {
        extend: function () {},
        useFrame: function () {},
        useLoader: function () { return {} },
        useThree: function () {
            return { }
        }
    }
});

import React from 'react';
import renderer from 'react-test-renderer';
import Ocean from './Ocean';

import {Vector3} from 'three'


const SUN_DIRECTION = new Vector3(1, 2, 3)
const PLANE_GEOMETRY_UUID = "123456789"


it('renders correctly', async () => {
    const tree = renderer
        .create(<Ocean sunDirection={SUN_DIRECTION}/>)
        .toJSON();
    tree.props.args[0].uuid = PLANE_GEOMETRY_UUID
    expect(tree).toMatchSnapshot();
});
