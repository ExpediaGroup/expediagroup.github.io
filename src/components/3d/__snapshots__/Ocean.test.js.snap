// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<water
  args={
    Array [
      Object {
        "height": 10000,
        "heightSegments": 1,
        "metadata": Object {
          "generator": "BufferGeometry.toJSON",
          "type": "BufferGeometry",
          "version": 4.5,
        },
        "type": "PlaneGeometry",
        "uuid": "123456789",
        "width": 10000,
        "widthSegments": 1,
      },
      Object {
        "distortionScale": 1,
        "fog": false,
        "format": undefined,
        "sunColor": 16777215,
        "sunDirection": Vector3 {
          "x": 1,
          "y": 2,
          "z": 3,
        },
        "textureHeight": 512,
        "textureWidth": 512,
        "waterColor": 1618851,
        "waterNormals": Object {
          "wrapS": 1000,
          "wrapT": 1000,
        },
      },
    ]
  }
  rotation-x={-1.5707963267948966}
/>
`;
